### PR TITLE
Competition API tests & enhancements

### PR DESCRIPTION
**Pull request by DennisSnijder**
# Context
A recent bug showed the need for tests on the competition endpoints.

# What has been done
- Competition creation API tests
- Moved the model validation to the competition service to remain consistent.
- Changed some Exceptions being thrown to return proper status codes.

### CODE REVIEW

File: server/src/api/modules/competitions/competition.service.js [hunk 1]
```diff
@@ -307,15 +309,23 @@ async function create(title, metric, startsAt, endsAt, groupId, groupVerificatio
...
+      title: sanitizedTitle,
+      metric: metric.toLowerCase(),
+      verificationCode,
+      verificationHash,
+      startsAt,
+      endsAt,
+      groupId
+    });
+  } catch (e) {
+    if(e instanceof ValidationError) {
```
Comment (DennisSnijder) [inline server/src/api/modules/competitions/competition.service.js:323]: @psikoi perhaps we should move the model validation to the service 🤔

**Comment by DennisSnijder** [comment]
@psikoi PR is ready! 😄 

image

File: server/src/api/modules/competitions/competition.service.js [hunk 2]
```diff
@@ -13,7 +13,7 @@ const {
 } = require('../../../database');
 const { durationBetween, isValidDate, isPast } = require('../../util/dates');
 const { generateVerification, verifyCode } = require('../../util/verification');
-const { BadRequestError } = require('../../errors');
+const { BadRequestError, NotFoundError } = require('../../errors');
```
Comment (psikoi) [inline server/src/api/modules/competitions/competition.service.js:16]: NotFound is 404, is this applicable to an id not being found?

**Comment by DennisSnijder** [comment]
@psikoi added some additional tests for adding participants, removing them and creating competitions as a group. Only thing I'm not 100% covering is all of the verificationCodes in each endpoint, which feels kinda repetetive since we're using the same logic for verifying that on pretty much all calls.

tldr; Ready for review once again

image

File: docs/src/configs/docs/competitions/endpoints.js [hunk 3]
```diff
@@ -236,7 +236,7 @@ export default [
       },
       {
         description: 'If startsAt OR endsAt is a past date.',
-        body: { message: 'Invalid dates: All start and end dates must be in the future.' }
+        body: { message: 'Start date must be before the end date.' }
```
Comment (psikoi) [inline docs/src/configs/docs/competitions/endpoints.js:239]: Is the previous message not relevant anymore? This doesn't seem to be just a grammar correction.

File: server/__tests__/integration/competition.test.js [hunk 4]
```diff
@@ -0,0 +1,512 @@
...
+      const response = await request.get(`/api/competitions`).send();
+
+      expect(response.status).toBe(200);
+      expect(response.body.length).toBe(3);
+
+      done();
+    });
+
+    test('Search competitions ( invalid status )', async done => {
+      const response = await request.get(`/api/competitions?status=invalid-status`).send();
```
Comment (psikoi) [inline server/__tests__/integration/competition.test.js:253]: It would be better to not hardcode the query params, check the Player search tests to see how you can do it differently ;)

File: server/__tests__/integration/competition.test.js [hunk 5]
```diff
@@ -0,0 +1,512 @@
...
+      done();
+    });
+
+    test('Search competitions ( with metric )', async done => {
+      const response = await request.get(`/api/competitions?metric=OvErAlL`).send();
+
+      expect(response.status).toBe(200);
+      expect(response.body.length).toBe(3);
+
+      await response.body.map(c => expect(c.metric).toBe('overall'));
```
Comment (psikoi) [inline server/__tests__/integration/competition.test.js:276]: why is there an `await` here?

File: server/__tests__/integration/competition.test.js [hunk 6]
```diff
@@ -0,0 +1,512 @@
...
+        verificationCode: TEST_DATA.minimal.verificationCode
+      });
+
+      expect(response.status).toBe(404);
+      expect(response.body.message).toBe('Competition of id 1234 was not found.');
+
+      done();
+    });
+
+    test('Do not remove participant ( No participants givens )', async done => {
```
Comment (psikoi) [inline server/__tests__/integration/competition.test.js:435]: "givens" 😆

File: server/__tests__/integration/competition.test.js [hunk 7]
```diff
@@ -0,0 +1,512 @@
...
+
+      expect(response.status).toBe(200);
+      expect(response.body.message).toBe(
+        `Successfully removed 1 participants from competition of id: ${TEST_DATA.minimal.id}.`
+      );
+
+      done();
+    });
+
+    test('Do not remove participant ( Not in participant list )', async done => {
```
Comment (psikoi) [inline server/__tests__/integration/competition.test.js:473]: Isn't this test repeated? line 447

File: docs/src/configs/docs/competitions/endpoints.js (hunk 3 above)
Comment (DennisSnijder) [reply docs/src/configs/docs/competitions/endpoints.js:239]: Yup, the invalid dates prefix was caused by the model validation, since most errors don't have such error prefix. I thought I'd remove it.

File: server/__tests__/integration/competition.test.js (hunk 7 above)
Comment (DennisSnijder) [reply server/__tests__/integration/competition.test.js]: yup! seems like I was copy pasting a bit too much haha

**Comment by DennisSnijder** [comment]
@psikoi remarks fixed!

File: docs/src/configs/docs/competitions/endpoints.js (hunk 3 above)
Comment (psikoi) [reply docs/src/configs/docs/competitions/endpoints.js:239]: No I mean the original meaning of the message, all start and end dates cannot be past dates?

Comment (DennisSnijder) [reply docs/src/configs/docs/competitions/endpoints.js:239]: Yeah, sounds kinda weird, what'd you suggest? 🤔
